source/agora/network/RPC.d

```diff
@@ -33,12 +33,91 @@ import vibe.core.connectionpool;
 import vibe.core.sync;
 import vibe.http.server : RejectConnectionPredicate;
 
+import std.conv : to;
+import std.socket : Address, AddressFamily, parseAddress;
 import std.traits;
 import core.stdc.stdio;
 import core.time;
 
 mixin AddLogger!();
 
+struct ProxyProtocol
+{
+    /// Protocol v2 signature
+    private immutable ubyte[12] v2_signature = [
+        13, 10, 13, 10, 0, 13, 10, 81, 85, 73, 84, 10
+    ];
+
+    /// Source address
+    Address src;
+
+    /// Destination address
+    Address dst;
+
+    ///
+    public this (TCPConnection stream) @trusted
+    {
+        ubyte[108] buffer;
+
+        stream.read(buffer[0..12]); // Read signature
+
+        ensure(buffer[0..12] != v2_signature,
+            "Proxy Protocol v2 is not yet supported");
+
+        // Read into buffer starting from `last_ix` until `ch`
+        uint readUntil (uint last_ix, char ch)
```

Comment:
Vibe.d really needs [this stuff](https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader.readline), noted